### ISSUE TITLE
Linux Kernel GPL and ZFS CDDL License clarifications in Support of Proposals in Issue #11357 and #13397 and #13349

### ISSUE THREAD

**Issue by jittygitty** [openzfs/zfs#13415]
### Describe the feature would like to see added to OpenZFS
This should be considered more like part of "DOCUMENTATION", and it is in "support" of the Proposals and/or Feature Requests made in Issues: #11357 and #13397 and #13349

If the proposals made in those issues are accepted, it may be beneficial to add the points presented here as justification/reasoning.

**NOTE THIS was just whipped up quick now, it is a major WIP, ie WORK IN PROGRESS and to be EDITED often to ADD/Modify/Etc.**

THE TITLE of this ISSUE was also supposed to INCLUDE SOMETHING along the lines of:
**Why the Future of ZFS on Linux is assured, regardless of any license.
And why Ubuntu wins legally even in their method of "distribution".**

### How will this feature improve OpenZFS?
There are many that have been needlessly scared away from ZFS on Linux by a lot of misinformation claiming license incompatibility.

I hope to clear up a lot of the misconceptions here and show how a relatively small license 'technicality' is a "moot" point in very large segments of users and use cases, and that in cases where the licenses leave some question marks and gray areas, that in those cases ZoL is on strong legal grounds, based on legal facts that have even recently been tested and proven in court. We shall touch on those.

Understanding license issues and unraveling the misinformation I hope will help openzfs continue to innovate and attract more talent and users of all kinds from home users to all other market segments. It should also help the leadership in accepting the proposals set forth in Issue #11357 and #13397 and #13349 etc. 
### Additional context

**Again this is a WIP, Work In Progress, I decided to write/update when I have time here and there etc, otherwise might not get done.**

First I think I should answer the question many people have had:

1. Why doesn't ORACLE re-license ZFS as Dual-License as other kernel modules are many times Dual Licensed as GPL+BSD/MIT/Apache.

Unfortunately it can't be licensed as Dual CDDL+GPL since unlike BSD/MIT/Apache, the CDDL does not allow you to change the license on original+modified files to a different license, like to GPL or to even custom/closed-source proprietary license.

**Please note that "license" compatibility issues are not just some Sun/Oracle invention, even Richard Stallman from the FSF can't make his own GPLv2 and GPLv3 to be "compatible", and clearly states GPLv2 and GPLv3 are "incompatible" licenses:**

"When we say that GPLv2 and GPLv3 are incompatible, it means there is no legal way to combine code under GPLv2 with code under GPLv3 in a single program. This is because both GPLv2 and GPLv3 are copyleft licenses: each of them says, “If you include code under this license in a larger program, the larger program must be under this license too.” There is no way to make them compatible."  https://www.gnu.org/licenses/rms-why-gplv3.en.html

(This is another good example of 'technical' complications that get in the way of the spirit of the license, he would have liked to make them compatible because their main intent was the same. And with CDDL and GPL in similar matter though they are both copyleft licenses and hence 'technically' incompatible, they both provide/protect the same freedoms (ie open source) 'intended' by the copyright holders and so a court can look at the intent and not count the 'technicality' in an infringement case.
Another example is with Peer-to-Peer downloading and BitTorrent which makes every downloader technically a "distributor", this "technicality" and the way it was addressed shows that such technicalities are sometimes inevitable problems that can and should be overlooked and given exceptions. Such realities weaken any case the FSF could bring against Ubuntu in the courts.)

If and when Oracle ZFS is no longer a real commercial competitor to OpenZFS and vice-versa, I believe Oracle would be open to relicensing ZFS on a Dual License, if the GPL version could be a version that had "patent protection" for them like the CDDL.

The CDDL (based on Mozilla Public License (MPL) v1.1) is a "copyleft" license just like GPL, though a bit more permissive than GPL since it isnt' as "viral", not everything it touches has to turn into GPL, With CDDL ('a bit similar' to LGPL) you only have to release the source of original files you modified and not your extra derivative works that work closely with those. But that's not I think the major issue, The issue for Oracle is that if they lose the CDDL, they lose its strong "PATENT" protection. You see the CDDL contrary to GPLv2 has a "patent grant" which is actually very nice. Since with GPLv2 you can actually get sued for using Linux by someone holding some patent who then claims Linux GPLv2 code is infringing on his patent. With CDDL you are given a patent grant, but the important part is that it is very "RECIPROCAL", that means that ZFS copyright holder/Oracle can't sue you over any patents they own for using ZFS, but in return, you can NOT sue Oracle for any "patent of yours" that ZFS might infringe upon or you lose your rights to ZFS license.

So that is the problem, one is that Oracle ZFS may still be a competitor, and two is the lack of good "patent protection" for Oracle with any of the current GPL licenses, even GPLv3 which did add a bit of patent retaliation I think. In fact, I think I found a thread with Sun employees mentioning even way back then, that the real reason for the CDDL was indeed patent protection and not as one sole employee (already trying to be the face/star of "Open Source") said (when no longer at Sun) that Sun had purposely made ZFS to be license incompatible with Linux. In fact I think she was publicly contradicted by even her supervisors or at least superiors at Sun etc.

What many upset with Sun/Oracle don't realize is that even if Oracle would accept more "limited" patent lawsuit protection from GPLv3 and wish to release ZFS as GPLv3 license, they can't because as mentioned GPLv3 is also incompatible with GPLv2.

(And LINUS has made it clear he never intends to upgrade Linux Kernel from GPLv2 to v3, because he frankly disagrees with the FSF in some big ways, he has no problem with Tivolization it seems, unlike the FSF. In fact I would say Linus has been misled by the above mentioned (refuted) claim that Sun intentionally chose CDDL so ZFS couldn't be on Linux and in fact by his emails its evident his concern was a lawsuit from Oracle for any such mixing, so it seems he would "not" take anyone to court over ZFS. )

The other issue, trying to avoid being as contagious/viral as GPL may have played a factor also, but declining with time and probably not much of a factor now. But sure, some wanted to be able to make some custom-proprietary closed-source products based on ZFS and with CDDL they could (and maybe the ZFS engineers and authors themselves wanted to do this at the time), but with GPL they can't.
(For example the Apache 2.0 license might have a good enough retaliatory patent grant as CDDL, though I haven't studied it yet, but I believe it was considered too liberal/permissive, they wanted something with at least weak 'copyleft' like in CDDL. But even if they were ok with Apache 2.0 they would still need to license as GPL also and as mentioned GPLv3 is still incompatible.)

So you see the combination of these two things is the reason why SUN left us with a license that can have a technical issue with GPL. Their engineers wanted the flexibility CDDL afforded them, and SUN wanted the patent protections, no conspiracy against open source at all, there just wasn't any good GPL compatible license that gave them what they needed as well, fair and simple.
It is not because SUN was evil and hated Open Source, to the contrary I think we have a lot to thank SUN for, few companies have done so much for open source as SUN. (OK even SGI gave us XFS, which might be awesome for Reflinks on a ZVOL?)

(I'm not in any way making Oracle the equivalent or successor of SUN "in that sense". RIP mysql, long live Maria...)

(And no it would not be smart for Oracle to come out and admit the importance of the patent protection to SUN or them in these matters, their lawyers would surely tell them not to say anything for good reasons. In fact their silence only reinforces the points I've made since it would have been safe and unproblematic for Oracle to confirm the (refuted) claim/rumor that they did not and do not want zfs on linux and that's the reason for the license being cddl. The silence on that hence confirms my points.)

3. Let's fast-forward and cut to the chase now. So why is Ubuntu and their daring "Distribution" of ZFS prebuilt binaries no problem?

There's "many" reasons that they win this battle legally, but for the USA courts let's put the best one first:

Because Ubuntu has "already" complied with the clearly stated "intent" of the Copyright holders, the kernel developers.

It's that simple. And it's one of the big reasons why they haven't been sued yet. Will explain more...

A. Though they are the most vocal, the Free Software Foundation (FSF) or Software Freedom Conservancy, (and I appreciate their other efforts against companies putting GPL into source-closed products and refusing to release code), do not have any "legal" standing to take someone to court for any supposed violation of GPLv2 kernel source code license because they might own the license, but they are not the "Copyright Holders" of the supposedly infringed upon Linux Kernel code.

In a recent related court battle, in Germany, they were I believe funding and supporting the Linux developer who was suing VMware, because again they have no legal standing to sue VMware themselves since they don't hold Linux Kernel copyright.

And why did they lose the court battle, and against a commercial company who admitted their software kernel contained some Linux kernel code? There's many reasons, but some big ones too, and if they lost in Germany/EU, they would lose worse in the USA where the laws make it even more difficult for open source to sue, since real damages must be proved and not just "hurt feelings" or some kind of "moral rights". In their particular case though, they seem to have come to the conclusion that in Germany, where I think "moral rights" can be taken in account and not just real damages incurred, that they were defeated because they did not come to court with "enough copyright holders" of the Linux kernel. Only one kernel developer just wasn't enough to represent the entire Linux Kernel from the German court's perspective. Especially since that "one" was not "LINUS"?

see: https://www.zdnet.com/article/linux-developer-abandons-vmware-lawsuit/
**"very difficult for individual Free Software developers to assert their rights alone."**
Meaning alone when they are trying to represent the Linux Kernel which has TONS of "copyright holders" with varying opinions.

And this is a big blow in itself since they can't easily fix it next time they try to go to court. Many kernel developers are perfectly ok even with CLOSED SOURCE kernel modules existing and being distributed as binaries, such as many graphics card "drivers".

And who can blame them if some of them make or made their daily bread by coding things that were "needed" by closed-source kernel modules etc. (Some might even be writing closed source modules while contributing to the Linux Kernel?)

When it comes to the "Export_Symbol_GPL" that was introduced to signify the "intent" of Linux kernel developers that such use of deep API "may mean" your kernel module "could" be considered a DERIVATIVE WORK. It does NOT mean that if you use those then your kernel module "IS" a derivative work. This is easily proven because Linus and other developers have agreed that for example that a kernel module that simply PORTED a file system from its "original operating system", would not be seen as a DERIVATIVE WORK, or in the "least" that it would not automatically be DERIVATIVE because of way it links or uses kernel.

This is a "huge" deal in a court room. It shows that EXPORT_SYMBOL_GPL is simply a "CODE COMMENT". You can do what you wish with it, it does NOT establish something as "Derivative" work. It also is NOT part of the license, you can't change/add to the GPL license in your code. One could say it might be "sort-of" like DRM rights management, that ironically I think the FSF and Kernel community is not too fond of right? But the best description is that it is a sort-of printed "CODE COMMENT" as a reminder/warning that IF your not simply porting something already existing on its own, like Andrew File System or ZFS, that IF you tie in too closely to the kernel code, that your kernel module "MIGHT" be considered as a "DERIVATIVE" work and if so would need to be open sourced and licensed GPL. You can read the very clear archived post from LINUS TORVALDS himself:

https://yarchive.net/comp/linux/gpl_modules.html

Now because several "EXPORT_SYMBOL" have been changed to "EXPORT_SYMBOL_GPL" and perhaps vice-versa it makes it even "more difficult" for anyone to use those to assert any license violation, since the license can't be changed in code etc. Not only that, but you can find Linux Kernel developers contradicting Linus' stance on license support intent by their discussions where they were saying perhaps ALL, EVERY single EXPORT should be made GPL at least initially, and ONLY after a lot of testing and usage and maturation "internally" with other GPL code would/should it be "opened" up to NON-GPL modules, which means they were talking of GPL-Only exports being used not as anything to do with License really but with technical management of bugs, to reduce bugs and effort because its very difficult for them to debug closed source kernel modules and fix any code in their own GPL code, that they wanted "EXPORT_SYMBOL_GPL" to be used for technical reasons to keep its usage to other GPL code first which is easier to debug, until they can fix most of their bugs, after which they would REMOVE GPL-only.

You can find discussions archived where they propose this and say flat out they are not against removing GPL-only later. You can even find in linux kernel source code a "EXPORT_SYMBOL_GPL" where it clearly tells you its really doesn't mean Gpl-only but its more of a DO NOT USE for technical reasons, unless you REALLY know what you're doing.

All of these things pretty much prove Export_Symbol_GPL is meaningless legally, especially for those who are adamant about abiding by all technicalities, since then they'd have to accept that "technically" the much touted GPL "Condom Works!".
https://lore.kernel.org/lkml/5fcc7fe6-9de3-cf34-075b-c17a28e85efa@tripleback.net/

To say the "shim" or gpl condom is a breach you'd have to play into "intent", because 'technically' it complies with GPL fine, And if you brush aside technicalities in favor of intent one way, then you can use the same intent to absolve CDDL technicality.

What does matter in US law is copyright holders' intent, and even here ZFS wins big time. Why?
Because read the Linux Kernel developers clear "intent": 

**"Kendall Bennett wrote:
>
> I have heard many people reference the fact that the although the Linux
> Kernel is under the GNU GPL license, that the code is licensed with an
> exception clause that says binary loadable modules do not have to be
> under the GPL.

Nope. No such exception exists." (LINUS TORVALDS responded)**

So their INTENT was for BINARY distributed modules to be GPL, they were against CLOSED-SOURCE modules and they repeated this stance against "CLOSED-SOURCE" proprietary modules consistently over the years. They never had a beef with Open Source modules! This is also why they can't go after ZFS, they would be proven hypocrites by their own prior clearly stated INTENT as copyright holders, that they just wanted whoever uses their code to also release their modifications etc.

Moreover, the CDDL actually lets the BINARY be GPL, so when the ZFS Kernel Module Itself BUILDS itself it can legally declare itself as GPL to the kernel module. The implication of this even for "distribution" is pretty clear, you can use any kernel API etc.

And now back to the biggest obstacle for a GPL lawsuit to win against Ubuntu, is because they've already complied with just about everything the Linux Developers, copyright holders have wanted:

1. ZFS on Linux is distributed in BINARY as GPL license.
3. ZFS on Linux is distributed in SOURCE as CDDL with a technical conflict with GPL but meeting the INTENT of the copyright holders. Even the REASON for the initial CREATION of the GPLv2 proves/supports that ZFS has "met" the copyright holders INTENT.
4. The ZFS on Linux doesn't include any non-copyleft files as far as I know, and includes info on the tooling and how to build etc, things that CDDL does not require but GPL does, but which Ubuntu has complied with even though CDDL didn't require.

A. So they have complied with the copyright holder's INTENT. 
B. By the copyright holders own statements, ZFS is likely NOT even a DERIVATIVE work, so they would NOT even have to comply and be GPL compatible in that case, it could even be CLOSED-SOURCE, because it existed for a different OS ie Solaris, just like Andrew File system that they gave an example of. So it should be considered professional largess to what lengths OpenZFS and Ubuntu have gone to be nice and comply as much as possible with the "Spirit of GPLv and Open Source" and the INTENT of the Linux Kernel copyright holders :)

**Oh and as I've said before, if you're openzfs or a company or anyone who doesn't even DISTRIBUTE openzfs prebuilt BINARIES along with LINUX, then NONE of these technical license issues matter, because if no distribution, the License has "no say at all".**

https://www.fsf.org/licensing/zfs-and-linux 
"Privately, You Can Do As You Like The GNU GPL has no substantive requirements about what you do in private; the GPL conditions apply when you make the work available to others." 

As I said in issue #11357 
https://github.com/openzfs/zfs/issues/11357#issuecomment-1118001842
"...
Now if the kernel Devs (Linus is much more level-headed but he's stuck with some fellows that went too extreme) would try to be more "just" and less hypocritical, they would "do away" with Export_Symbol_GPL and use Export_Symbol_COPYLEFT so that they recognize their original intent which is that they want to avoid closed-source binary blobs and have modified code be released under open source. This would have been a much more civil and just way to proceed versus the current license fiasco.

Very strong arguments can be made that OpenZFS can definitely qualify as a "non-derivative" work based on Linus' own past well archived remarks. So it is very hypocritical to treat OpenZFS or Ubuntu which complies with basically the very "intent" and reasons that Linus went with the GPL in 1992, to treat them just as bad, or actually "worse" than CLOSED-Source Binary BLOBS.
(And by worse I mean sometimes things got into the kernel which actually "closed source" drivers needed or benefited from, imagine if OpenZFS asked the kernel devs to put some new code (ie compatible with GPL) into the kernel to 'help' zfs module?)

If you read some of the email threads among the Kernel Devs, some are even proposing to somehow code into the kernel some DRM digital rights management (which kernel devs and the FSF (creators of GPLv2) have decried for years as horrible), so that they could somehow try to ENFORCE the GPL-Only "Exports" if some module tried to use a shim or other method to use indirectly.
They have been also contemplating trying to make "ALL" Exports to be GPL-only, to "change" all remaining ones to GPL-only.

I thought they were "champions" of copyleft "Open Source" and the spirit of "freedom", if that were the case why would they not "differentiate" between "CLOSED SOURCE" and try to punish other "copyleft" licenses of coders like them who wanted to protect the SAME freedoms for their own users but simply chose a "different" but still COPYLEFT license? Are they champions of some "magic" letters, ie G and P and L? Is there something special about those letters and no other copyleft license is worthy?
Even the FSF couldn't make their own GPLv3 to be "compatible" with the GPLv2 they created, should GPLv3 then be treated with just as much disdain as CLOSED-SOURCE because of that? (sorry I should of put this in https://github.com/openzfs/zfs/issues/13415 instead of here I guess.)
"
"
**Question: Would someone be willing to do a PULL-REQUEST to Linux Kernel repo with a patch of: EXPORT_SYMBOL_GPL to EXPORT_SYMBOL_COPYLEFT? Would be interesting to see the response from "everyone" including Kernel devs and Linus himself :)**
" https://github.com/openzfs/zfs/issues/11357#issuecomment-1118003748

**In fact, even Richard Stallman and the FSF who have always been more extreme than Linus or any kernel devs would have a lot to gain from that EXPORT_SYMBOL_COPYLEFT patch I proposed, since it would allow anyone to finally release kernel modules using the GPLv3 which is "technically" incompatible with GPLv2 just like CDDL, since copyleft licenses are like male betta fish.
(And FSF has been trying very hard to increase "adoption" of GPLv3 since they fixed "Tivolization" loophole through it etc...)**

There's much more to be added but, other things call...if you enjoyed the above and want more though do say...thx!

**Comment by gmelikov** [comment]
https://openzfs.github.io/openzfs-docs/License.html

**Comment by zfsbot** [comment]
the project shouldn't suggest or even document how to violate licenses.

also, there's a lot of fearmongering in your own proposal. how do missing GPL symbols impact performance or stability at all? WHY do we care about this change?

**Comment by jittygitty** [comment]
@zfsbot Not sure if you read this or my linked issues, since:

1. I never proposed that OpenZFS should do anything to "violate" any license.
2. This issue was specifically to undo "misconceptions" about licensing, to clarify questions many have.
3. GPL symbols can indeed impact performance or stability which was reason I posted #11357

You can look at https://github.com/openzfs/zfs/issues/8259 to see a clear example of "regression" caused by those not-so-harmless GPL symbol exports. My proposal was very simple and very absolutely legal:

**ALWAYS give us the end users the ability to choose to "USE" GPL-Only exports.
But this also means to "actually CODE FEATURES in a way to be ABLE TO optionally USE THEM when its advantageous".**

I'm glad that @behlendorf closed #11357 after he answered my question because it was turning into misguided attacks on me.

By the way, I would usually agree with you on what you said in there, ie:
"raising the ire of the upstream kernel developers is not a good idea. it has not worked out well for anyone so far."

But if you read just a few Linux kernel email threads, you'll see that boat has kinda sailed already.
https://lore.kernel.org/lkml/20190123155850.GB8721@amd/

As I've made clear in my comments on here, I "still" wish to be friends with the kernel guys, and whatever I do, the method will be either tongue-in-cheek funny while attracting a secondary serious and cordial friendly discussion, or via side-channels with someone who's already a contributing linux kernel developer who already disagrees with some of the too extreme views there.

But yes, I would reserve the right to do a pull-request/suggestion of exactly what I said, ie making a distinction between "CLOSED-SOURCE BINARY BLOBS" and honest to goodness fully "open source" projects like OpenZFS and others who just happen to be the SAME as the GPL in the open-source rights and freedoms they convey to users but simply by diff License.
Yet, "if" I do, I would not do it with this username or make any references to zfs actually, because its for "all copyleft", not zfs.

Currently, truly open-source projects, just as open source as GPL ones, are being treated as some sort of evil Closed-Source.

**As I mentioned, even if OpenZFS were not CDDL but GPLv3, we would have the "same exact" discussions we are having now.**

**Comment by zfsbot** [comment]
kernel symbols come and go. it is fact of life that a new kernel release will change things. this is a normal regression and fact of being out of tree module. there is no performance impact of not using GPL symbols. as Brian said, not using them sometimes makes code more maintainable as a matter of being more portable.

i wish you could see that your actions themselves feel like a misguided attack on the project. even now you "reserve 'the right'" to make a submission to a project to change their license policy based on your _own_ ideology.

GPLv3 isn't relevant, and the LKML thread you linked is _precisely_ the one i was referring to when Brian and others told anyone who wants to contribute to this project, to stop doing things like this. no matter whether its under your name, or referencing OpenZFS, you've essentially put a spectre of distrust on this whole community now by saying you will even do it underhandedly with a pseudonym. 

honestly, i wouldn't be surprised if you were banned from both places, just like the University of Wisconsin.

**Comment by jittygitty** [comment]
@zfsbot Obviously you are still not understanding, maybe you refuse to read what I've written, which tries hard to clarify things.

You just said:
"i wish you could see that your actions themselves feel like a misguided attack on the project"

Which project do you think I'm "attacking"? And why/how?

I explained quite well I think the very sound reasoning why I reserved the right, and in a nice way not mean-spirited, to make a submission like everyone else can and does with no guarantee of acceptance. What you are wrong about is that my request would not at all have to imply they must "change their license policy", not at all. The pull-request is for CODE that was 'meant' as reminder to "CLOSED-SOURCE BINARY BLOBS" that they risk becoming derivative works and need to release their "sources".

Also, even if it were about "license" ideology and not as I explained above, the ideology matches Linus' own when he went for GPL back in 1992, to release open source and have modifications also be released open source without encumbering the users.

**And hate to be forced to repeat myself, but I said "if" I would push some request to recognize other copyleft license like GPLv3 should be treated "better" than CLOSED-SOURCE Binary Blobs, I would "NOT" mention anything about zfs,openzfs, or "CDDL".**

And I'll quote myself yet again exactly what I said in previous post:
"
**As I've made clear in my comments on here, I "still" wish to be friends with the kernel guys, and whatever I do, the method will be either tongue-in-cheek funny while attracting a secondary serious and cordial friendly discussion, or via side-channels with someone who's already a contributing linux kernel developer who already disagrees with some of the too extreme views there.

But yes, I would reserve the right to do a pull-request/suggestion of exactly what I said, ie making a distinction between "CLOSED-SOURCE BINARY BLOBS" and honest to goodness fully "open source" projects like OpenZFS and others who just happen to be the SAME as the GPL in the open-source rights and freedoms they convey to users but simply by diff License.
Yet, "if" I do, I would not do it with this username or make any references to zfs actually, because its for "all copyleft", not zfs.

Currently, truly open-source projects, just as open source as GPL ones, are being treated as some sort of evil Closed-Source.

As I mentioned, even if OpenZFS were not CDDL but GPLv3, we would have the "same exact" discussions we are having now.**
"

When did Brian say that losing a GPL export can't cause performance issues? You weren't aware of the ton of "performance" complaints when openzfs lost access to fpu SIMD etc? Of course using about nothing from the kernel would make everything more portable, but obviously sometimes there's advantages, you really completely deny that? Where did Brian say there's never a performance or other advantage in using GPL exports? If that were true, wouldn't that prove the Kernel devs to be clueless?
Since obviously they must think there's sometimes quite an advantage to be able to use them, so are they as wrong as I am?

Also, tell me why isn't GPLv3 relevant to this discussion? 

(It is just as incompatible as CDDL to GPLv2, as I've said, and clearly stated by the FSF as well.)

If the Kernel Devls refused such a cordial pull-rquest they could at least be asked to fix their "TYPO" and correct: "Export_Symbol_GPL"
to
"Export_Symbol_GPLv2"

**Comment by zfsbot** [comment]
i'm not reading these massive wall of text posts.

**Comment by gmelikov** [comment]
Honestly, I cannot summarize your proposals in this issue's format.

We already have documentation about license https://openzfs.github.io/openzfs-docs/License.html , feel free to contribute (see "Edit on GitHub" link in top right corner of it), where you can clarify problematic pieces and we can easily review. I'll close this issue.
